wireapp/wire-ios-mono
# feat: NFCG error alert presenting WPB-2293

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-2293" title="WPB-2293" target="_blank"><img alt="Story" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10815?size=medium" />WPB-2293</a>  [iOS] Conversation creation errors on non-fully connected graph
  </td></table>
  <br />
 




















----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

There is a need to present an alert for not fully connected graphs when from certain backends cannot federate between each other. This requires to connect to new backend response while creating a conversation.

### Issues

- Creation flow issue
- General flow of error handling
- Error handling
- Delegates

### Causes (Optional)

1. Creation flow issue - was first dismissing conversation creation and opening created conversation and then updating the conversation with users and settings. In case of any errors the conversation was created anyway with unmodified settings or users list. Then the popup was presented on the already opened conversation, and if error that requires conversation to be deleted was present, then the error alert was displayed and the conversation was closed. While investigating it turned out that such alerts were dismissed immediately after presenting due to closing of the conversation view, causing alerts to be unreadable due to auto-closing.
2. General flow of error handling - Errors were handled after closing creator and opening conversation. New NFCG alert should allow users to edit participants list manually or closing the creator. That means that errors should be handled prior to closing creator and opening conversation.
3. Error handling - it is performed by usage of notifications. This complicates the flow as there is no immediate information about success or failure of conversation creation. The observers for conversation updating related issues were setup in `ConversationListContentController` instead of the creator, which complicates the whole flow.
4. Conversations can be created from multiple places and instead of being created from `ConversationCreationController` they were created through multiple layers of delegates which leads to high logic complexity and no unification.

### Solutions

1. Modifying the flow:
a. removed the complex delegates passing
b. handling errors on `ConversationCreationController` instead of `ConversationListContentController`
c. first the conversation is created & updated, then the creator is closed and conversation opened if needed instead of first closing creator, opening conversation and then creating and updating it
d. Introduced `GroupConversationCreationCoordinator` for managing creation flow and actions
e. Added loading indicator while conversation is being created
f. Simplified flow
2. Modifying error handling:
a. `GroupConversationCreationCoordinator` listens to notifications instead of `ConversationListContentController`
b. Error, success and all requested actions are returned from `GroupConversationCreationCoordinator` to the UI
3. NFCG:
a. Connected to backend response regarding non-federating backends and handling it
b. Modified strategy accordingly
c. Introduced notifications to pass failures and success information from strategies back to `GroupConversationCreationCoordinator`
d. Presenting a new alert with appropriate message and possible actions

### Testing

#### Test Coverage (Optional)


#### How to Test

Create group conversation on backend _A_ either through profile or contacts. In creation wizard, add federated users from backends _B_ and _C_ to the conversation along with users from _A_. If backends _B_ and _C_ does not federate with each other, then the NFCG alert should be presented.

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
